### PR TITLE
Update to angular 10

### PR DESCRIPTION
**Pull request by mocleiri**
Needed to apply a trick inside of app.module.ts to prevent tree shaking of the example components when building in prod mode.

See: https://github.com/angular/angular/issues/35314#issuecomment-584821399

Added a startup command to run ng serve but in prod mode which was used
to verify the examples working after the tree shaking fix.

Also had to exclude the path to the example javascript as ivy was finding it and breaking since its not typescript.

### CODE REVIEW

File: angular.json [hunk 1]
```diff
@@ -13,6 +13,12 @@
         "build": {
           "builder": "@angular-devkit/build-angular:browser",
           "options": {
+            "allowedCommonJsDependencies": [
+              "lodash",
+              "util",
+              "xlsx",
+              "file-saver"
+            ],
```
Comment (mocleiri) [inline angular.json:21]: removes complaints about common js modules

File: angular.json [hunk 2]
```diff
@@ -22,7 +28,8 @@
             "aot": true,
             "assets": [
               "src/favicon.ico",
-              "src/assets"
+              "src/assets/logos",
+              "src/assets/md-files"
             ],
```
Comment (mocleiri) [inline angular.json:33]: exclude **src/assets/example-sources** so it doesn't get processed by typescript/javascript compiler.

File: package.json [hunk 3]
```diff
@@ -11,58 +11,60 @@
   "scripts": {
     "ng": "ng",
     "start": "ng serve",
+    "start-prod": "ng serve --prod --aot=true --optimization=true",
```
Comment (mocleiri) [inline package.json:14]: I used this to first see that the examples were missing and then to see that they were present when I found out how to refer to them in the AppModule which prevents them being tree shaken out.

File: src/app/app.module.ts [hunk 4]
```diff
@@ -39,10 +39,8 @@ export function markedOptions(): MarkedOptions {
 
   return {
     gfm: true,
-    tables: true,
     breaks: false,
     pedantic: false,
-    sanitize: false,
     smartLists: true,
     smartypants: false,
   };
```
Comment (mocleiri) [inline src/app/app.module.ts:46]: I think this is API changes in the object where tables and sanitize weren't in MarkedOptions anymore.

File: src/app/app.module.ts [hunk 5]
```diff
@@ -104,4 +102,17 @@ export function markedOptions(): MarkedOptions {
...
+    ArrayFilterComponent,
+    BriefExporterComponent,
+    CustomColumnFilterComponent,
+    CustomExporterComponent,
+    SimpleFilterComponent,
+    PropertyOptionsComponent,
+    SelectionExporterComponent
+
+  ]
+}
```
Comment (mocleiri) [inline src/app/app.module.ts:118]: This is the work around to prevent the example components to be tree shaken out of the production build code.  

Essentially the old way of listing them as part of the @NgModule declaration works differently in ivy.

File: tsconfig.json [hunk 6]
```diff
@@ -50,6 +50,6 @@
   "angularCompilerOptions": {
     "fullTemplateTypeCheck": true,
     "strictInjectionParameters": true,
-    "enableIvy": false
+    "enableIvy": true
```
Comment (ccosmincc) [inline tsconfig.json:53]: Angular team does not yet recommend enabling this ( https://angular.io/guide/creating-libraries#publishing-your-library ). What is the reason behind this decision?

File: package.json (hunk 3 above)
Comment (ccosmincc) [reply package.json:14]: Since this was meant only for debugging purposes, do we see any use in still keeping it after the debug was finished?

File: package.json [hunk 7]
```diff
@@ -11,58 +11,60 @@
...
     "build": "ng build",
     "test": "ng test",
     "lint": "ng lint",
     "e2e": "ng e2e"
   },
   "private": true,
+  "peerDependencies": {},
```
Comment (ccosmincc) [inline package.json:21]: Why is this needed?

File: angular.json (hunk 2 above)
Comment (ccosmincc) [reply angular.json:33]: Instead of explicitly enumerating all the sub-folders (which will cause the need to always update this whenever a new folder will be added to `src/assets`), why not just say: take everything from `src/assets` folder, except `example-sources` folder? Something like this:
`{
                "glob": "**/*",
                "input": "src/assets",
                "output": "/assets",
                "ignore": ["examples-sources/*"]
}`

Comment (mocleiri) [reply angular.json:33]: yes, that would be better.  

Does the {} expression directly replace the "src/assets" line (line 25)?

File: package.json (hunk 7 above)
Comment (mocleiri) [reply package.json:21]: can be removed.

File: package.json (hunk 3 above)
Comment (mocleiri) [reply package.json:14]: you're right I think this can be removed.

File: tsconfig.json (hunk 6 above)
Comment (mocleiri) [reply tsconfig.json:53]: That recommendation is for libraries written in angular 9 where ivy was optional.  

I was thinking that if you are using angular 9 and the ng-material-extensions library then you are going to be consuming a 9.0.x library version and that if you are using angular 10 then you would use the 10.0.x version (this pull request) so that that guidance would not apply here.

My angular 10 project using ng-material-extensions version 9 doesn't build right with aot on.  I have to test it with this pull request build to see if it works and as part of that I can see if this switch is important or not.

File: package.json (hunk 7 above)
Comment (mocleiri) [reply package.json:21]: this can be removed.

File: angular.json (hunk 2 above)
Comment (ccosmincc) [reply angular.json:33]: Correct. Assets array should look like this:
`
"assets": [
              "src/favicon.ico",
               { "glob": "**/*", "input": "src/assets", "output": "/assets", "ignore": ["examples-sources/*"] }
            ]
`

**Comment by rammohan201** [comment]
when will this PR be merged for using?

**Comment by digitaldrummerj** [comment]
@HalitTalha @ccosmincc @jdpnielsen any updates on when you will have a chance to review and possibly merge this?

**Comment by HalitTalha** [comment]
Thanks for the effort. It appears except some unidentified rare cases with aot the ng9 built packages are working fine up to the latest ng version. I feel confident about this since I got the chance to test and also seen some enterprise grade applications migrated to ng 10 without any issue. If you think otherwise, any feedback is welcome.
We'll go with updating only peer dependency versions for now to avoid the warnings during package installations.

There is an ongoing major upgrade we are working on, it will bring some performance improvements and remove cdk datasource page based implementation.
